Fake PR, just for reviewing purposes

lsago opened the pull request.

JHogenboom committed:
initial commit

JHogenboom committed:
correct algorithm name

JHogenboom committed:
correct Dockerfile

JHogenboom committed:
correct dependencies

JHogenboom committed:
update permissions for docs

JHogenboom committed:
correct numerical aggregation

JHogenboom committed:
update algorithm base version

JHogenboom committed:
correct handling of categorical outliers

JHogenboom committed:
initial commit sparql implementation

JHogenboom committed:
improve sparql implementation - 'functionised' and accredit helpline

JHogenboom committed:
correct handling of missing data

JHogenboom committed:
include organisation id in partial for traceability

JHogenboom committed:
update SPARQL implementation

JHogenboom committed:
fetch names instead of ids

JHogenboom committed:
update algorithm_store.json

JHogenboom committed:
create a separate ID as URI identifiers are not reliable

JHogenboom committed:
prevent errors for variables that are not found in the triplestore

JHogenboom committed:
include snomed

JHogenboom committed:
correct algorithm_store.json typo

JHogenboom committed:
improve handling of value classes and prefixes

JHogenboom committed:
remove checklist

JHogenboom committed:
update algorithm_store.json to vantage6 v3.10 standards

JHogenboom committed:
improve handling mixed datatypes, solving issue #2

JHogenboom committed:
improve handling of data graphs with varying length

JHogenboom committed:
adapt inlier input check to JSON item

JHogenboom committed:
prevent duplicate class associations

JHogenboom committed:
correct SPARQL query

JHogenboom committed:
update Dockerfile and algorithm base image

JHogenboom committed:
update LICENCE

JHogenboom committed:
initial plan of using general libraries

JHogenboom committed:
update general implementation

JHogenboom committed:
Merge pull request #6 from STRONGAYA/revamped-version

Updated implementation using v6-tools-general

Copilot committed:
Initial plan

Copilot committed:
Initial analysis and planning for comprehensive testing framework

Co-authored-by: JHogenboom <106407313+JHogenboom@users.noreply.github.com>

Copilot committed:
Implement comprehensive testing framework and production-grade CI/CD

Co-authored-by: JHogenboom <106407313+JHogenboom@users.noreply.github.com>

Copilot committed:
Add framework validation and standalone testing capabilities

Co-authored-by: JHogenboom <106407313+JHogenboom@users.noreply.github.com>

Copilot committed:
Consolidate workflows and restructure tests based on feedback

Co-authored-by: JHogenboom <106407313+JHogenboom@users.noreply.github.com>

Copilot committed:
Update README with tested Vantage6 versions and streamlined workflow documentation

Co-authored-by: JHogenboom <106407313+JHogenboom@users.noreply.github.com>

JHogenboom committed:
update image building and developer network setup

JHogenboom committed:
rearrange file structure

JHogenboom committed:
correct path for basic function tests

JHogenboom committed:
rearrange developer network and image building for session usage

JHogenboom committed:
start testing with more recent version

JHogenboom committed:
update test procedure with configurable vantage6 workflow

JHogenboom committed:
clean up node config

JHogenboom committed:
remove creatures_of_enceladus.csv diet

JHogenboom committed:
clean up various redundant tests and configurations

JHogenboom committed:
resolve various tests failures

JHogenboom committed:
update various test configurations

JHogenboom committed:
simplify docker ip fetching

JHogenboom committed:
fix datatypes and include IP logging

JHogenboom committed:
include http in docker gateway check

JHogenboom committed:
include retry mechanism for node connections and resolve deprecation error

JHogenboom committed:
fetch first 100 lines for connection test

JHogenboom committed:
update algorithm implementation to resolve various test cases

JHogenboom committed:
update testing setup

Copilot committed:
Fix failing tests and implement TODOs as requested

Co-authored-by: JHogenboom <106407313+JHogenboom@users.noreply.github.com>

JHogenboom committed:
update misc and README.md

JHogenboom committed:
update test config and method control

JHogenboom committed:
include rdf implementation

JHogenboom committed:
update equivalency testing

JHogenboom committed:
update README.md

JHogenboom committed:
update documentation

JHogenboom committed:
fix inlier related failure

JHogenboom committed:
switch to localhost in the non-dockerised client for robustness

JHogenboom committed:
update algorithm store template

JHogenboom committed:
use black and flake8

Copilot committed:
Fix test failures and improve inlier handling

Co-authored-by: JHogenboom <106407313+JHogenboom@users.noreply.github.com>

JHogenboom committed:
abandon lazy partial retrieval function

JHogenboom committed:
Merge remote-tracking branch 'origin/copilot/fix-7' into copilot/fix-7

Copilot committed:
Fix inlier test configurations and improve count validation logic

Co-authored-by: JHogenboom <106407313+JHogenboom@users.noreply.github.com>

JHogenboom committed:
update equivalence determination

Copilot committed:
Fix inlier test validation logic and enhance categorical value count verification

Co-authored-by: JHogenboom <106407313+JHogenboom@users.noreply.github.com>

JHogenboom committed:
update inlier handling and enforce sample size threshold on the result

JHogenboom committed:
update docstring

JHogenboom committed:
update docstring

JHogenboom committed:
update linting

JHogenboom committed:
remove coverage - as pytest cannot assess coverage in this Dockerised setup

JHogenboom committed:
broaden matrix to see compatability

JHogenboom committed:
reduce vantage6 matrix due to certain test incompatibility

JHogenboom committed:
Merge pull request #8 from STRONGAYA/copilot/fix-7

Introduce testing setup and associated test failures

JHogenboom committed:
bump rdf tools version

JHogenboom committed:
update release workflow with renewed algorithm

JHogenboom committed:
Update STRONG AYA Algorithm Guideline badge version

JHogenboom committed:
update status badge

JHogenboom committed:
Merge remote-tracking branch 'origin/main'

JHogenboom committed:
update handling of variables_to_describe and partially-existing inputs

JHogenboom committed:
warn for adjusted privacy threshold violations, rather than raising an error

JHogenboom committed:
correct linting

JHogenboom committed:
update sample size check of results, raise error when nothing is left, warn otherwise

JHogenboom committed:
update sample size check of results, raise error when nothing is left, warn otherwise

bartvanb commented on v6-descriptive-statistics/central.py:
```diff
@@ -0,0 +1,139 @@
+    # Check if the users' input structure is correct
+    if not check_input_structure(variables_to_describe):
+        raise UserInputError(
+            "Algorithm input is incorrect. Please check the algorithm input."
```
This message would ideally be a bit more helpful in what is wrong with the input

JHogenboom committed:
include DOI in README.md

JHogenboom committed:
bump tools' version

JHogenboom committed:
bump algorithm version

bartvanb commented on README.md:
```diff
@@ -1,2 +1,146 @@
+- Frequency counts for categorical variables
+
+A more detailed description of the algorithm and how to use it can be found in the algorithm's
+[Wiki](https://github.com/STRONGAYA/v6-descriptive-statistics/wiki) and the `/docs` directory.
```
I just checked the wiki and I don't see much info there. Perhaps the import went wrong?

bartvanb commented on algorithm_store.json:
```diff
@@ -0,0 +1,161 @@
+{
+  "description": "Algorithm to retrieve descriptive statistics. For categorical data this include value, outlier, and missing values counts. For numerical data this includes minimum, maximum, counts (total number of values), missing value counts, outlier counts, deviation sum, sum, mean, and deviation.",
+  "documentation_url": "https://github.com/STRONGAYA/v6-descriptive-statistics/wiki",
+  "image": null,
```
just wondering why this is `null`? Have you tried submitting this to a store?

bartvanb commented on setup.py:
```diff
@@ -0,0 +1,28 @@
+    packages=find_packages(),
+    python_requires=">=3.10",
+    install_requires=[
+        "vantage6-algorithm-tools",
```
Might be useful to add `<5.0` here so that it still works in v4 once v5 is out

Same in requirements.txt

bartvanb replied:
important

JHogenboom replied:
Resolved in bc6f4ef5d66e63646f28eb072cc6f33d42883b05 with switch to `pyproject.toml` from `requirements.txt` and `setup.py`

`pyproject.toml`
```toml
dependencies = [
    "vantage6-algorithm-tools==4.15.1",
    "pandas>=2.2.0",
    "requests",
    "numpy",
    "vantage6-strongaya-general @ git+https://github.com/STRONGAYA/v6-tools-general.git@v1.0.3",
    "vantage6-strongaya-rdf @ git+https://github.com/STRONGAYA/v6-tools-rdf.git@v1.0.1",
]
```

bartvanb commented on docs/v6-descriptive-statistics/usage.rst:
```diff
@@ -0,0 +1,71 @@
+---------------
+The input arguments for the central function consist of:
+
+- variables_to_describe (dict): Dictionary of variables to describe;
```
What should this dict look like? 

There's an example below, but are that all the possibilities listed there or are there more options? It would be good if the docs are explicit on that

bartvanb commented on docs/v6-descriptive-statistics/usage.rst:
```diff
@@ -0,0 +1,71 @@
+
+- variables_to_describe (dict): Dictionary of variables to describe;
+for categorical outliers: we assume that anything that is not in the inlier tuple is an outlier, and for numerical outliers the tuple should consist of two values, anything that is not within the range of the tuple is then considered an outlier.
+- variables_to_stratify (dict, optional): Dictionary of variables to stratify
```
same here

bartvanb commented on docs/v6-descriptive-statistics/usage.rst:
```diff
@@ -0,0 +1,71 @@
+- variables_to_describe (dict): Dictionary of variables to describe;
+for categorical outliers: we assume that anything that is not in the inlier tuple is an outlier, and for numerical outliers the tuple should consist of two values, anything that is not within the range of the tuple is then considered an outlier.
+- variables_to_stratify (dict, optional): Dictionary of variables to stratify
+- organization_ids (list, optional): A list of organisation IDs that participate in the collaboration and you wish to run the algorithm on.
```
What happens if you don't specify it?

bartvanb commented on docs/v6-descriptive-statistics/privacy.rst:
```diff
@@ -0,0 +1,91 @@
+**Risk analysis**:
+The algorithm does not produce a model and therefore a watermark attack thereon is not perceived as a privacy risk.
+
+.. TODO verify whether these definitions are correct.
```
just alerting you to open TODO

bartvanb commented on v6-descriptive-statistics/central.py:
```diff
@@ -0,0 +1,139 @@
+                                                                 {"Gender": {"datatype": "categorical",
+                                                                             "inliers": ("M", "F", "X")},
+                                                                  "Age": {"datatype": "numerical",
+                                                                          "inliers": (15, 39)}},
```
I think I noted this elsewhere as well. An example is not enough documentation - all possible options should be available, or a link to where all options can be seen. Otherwise it is not clear how to use the algorithm.

Same for the argument below.

bartvanb commented on v6-descriptive-statistics/central.py:
```diff
@@ -0,0 +1,139 @@
+    # Compute the aggregate of the aggregate-adjusted deviation and include it in the general statistics
+    results = compute_aggregate_adjusted_deviation(
+        results_deviation, results_general_statistics
+    )
```
IIRC you mentioned that privacy checks would be called here in the main function, but I don't see it. What am I missing?

bartvanb replied:
Ah right, it's done already in the partials of course

bartvanb commented on v6-descriptive-statistics/miscellaneous.py:
```diff
@@ -0,0 +1,243 @@
+                if not all(isinstance(x, str) for x in inliers):
+                    return False
+
+    return True
```
I think there are several libraries (e.g. marshmallow, pydantic) that make this kind of schema checking easier to implement and easier to read. It would be nice to use these.

bartvanb commented on v6-descriptive-statistics/miscellaneous.py:
```diff
@@ -0,0 +1,243 @@
+        PrivacyThresholdViolation: If privacy threshold violations result in no remaining statistics
+    """
+    # Retrieve the sample size threshold
+    sample_size_threshold = get_env_var("SAMPLE_SIZE_THRESHOLD")
```
Is it part of the documentation that you can set this env var and it will update the threshold?

bartvanb replied:
important

JHogenboom replied:
Resolved in cc40169ce1471b3bbd1c204e1b7f7987662127fb and communication

`README.md`
```markdown
For usage, we recommend that within your collaboration, every node should configure the `SAMPLE_SIZE_THRESHOLD` 
environment variable to a value that is you have reached a consensus on.
If not specified, the default value is that the working dataframe should at least contain 10 records.
```

bartvanb commented on v6-descriptive-statistics/miscellaneous.py:
```diff
@@ -0,0 +1,243 @@
+    try:
+        sample_size_threshold = int(sample_size_threshold)
+    except TypeError:
+        sample_size_threshold = 10
```
It might be nice to print a warning in the node logs that the threshold from the node config is not applied. Note also that the `get_env_var` function has optional args to provide default and require it to be returned as int (or `EnvironmentVariableError` is raise)

bartvanb replied:
important

JHogenboom replied:
Resolved in cc40169ce1471b3bbd1c204e1b7f7987662127fb

`v6-descriptive-statistics/miscellaneous.py`
```python
sample_size_threshold = get_env_var("SAMPLE_SIZE_THRESHOLD", default=10, as_type="int")
```

bartvanb commented on docs/v6-descriptive-statistics/validation.rst:
```diff
@@ -0,0 +1,19 @@
+Validation
+==========
+Empirical testing
+====================
```
I'd expect an empty line between different headers (otherwise it's displayed weirdly I think) and also, when using the same characters for the title underline (in this case `=`), you will get the same level header, so better to use another?

bartvanb commented on v6-descriptive-statistics/miscellaneous.py:
```diff
@@ -0,0 +1,243 @@
+                )
+                filtered_result["categorical_general_partial_statistics"] = (
+                    combined_categorical_df.to_json()
+                )
```
I see several long if's here. To make it more readable, it could be refactored to have code like

```python
def check_and_enforce_threshold():

    threshold = _get_threshold()
    
    _check_categorical()

    _check_numerical()

    _other_stuff_with_a_better_name_than_this()
```

I guess it's OK as is but for maintainability of this code for the future, this would be preferable

bartvanb commented on v6-descriptive-statistics/partial.py:
```diff
@@ -0,0 +1,208 @@
+                                                                 {"Gender": {"datatype": "categorical",
+                                                                             "inliers": ("M", "F", "X")},
+                                                                  "Age": {"datatype": "numerical",
+                                                                          "inliers": (15, 39)}},
```
I mentioned this elsewhere I think, but it would be better if you describe what the allowed structure of the dict is instead of only giving an example (giving example is good but it should not be just that)

bartvanb replied:
Also, I think requiring such a rather difficult structure as user input is not making it easy for users to run this algorithm themselves. This is fine if we hide the algorithm behind an interface (e.g. a UI), is that the case?
I'm pretty sure this is difficult to run in the vantage6 UI, we may consider adding support there for such cases if an easier structure is not possible.

bartvanb commented on v6-descriptive-statistics/partial.py:
```diff
@@ -0,0 +1,208 @@
+    else:
+        variables_to_analyse = list(variables_to_describe.keys())
+
+    # Retrieve RDF/SPARQL data if its use is indicated in the data - suboptimal solution, to be improved in the future
```
create issue to log this?

bartvanb commented on v6-descriptive-statistics/miscellaneous.py:
```diff
@@ -0,0 +1,243 @@
+                    privacy_violations.append(f"categorical_violation_{variable_name}")
+                    continue  # Skip this variable entirely
+
+                # Keep valid data rows and all metadata rows (nan/outliers are always allowed)
```
If I understand correctly, the results will show that there is exactly 1 outlier? 
I guess for this algorithm that is fine, as there is probably no way to find out what the value of the outlier is (?), but this might open it up (slightly) to find out what the value might be by running other algorithms.

bartvanb commented on v6-descriptive-statistics/miscellaneous.py:
```diff
@@ -0,0 +1,243 @@
+                # Find the count row for this variable
+                count_row = variable_rows[variable_rows["statistic"] == "count"]
+
+                if not count_row.empty:
```
if count row is empty, the row is added to the result - to me it seems better at first sight to NOT add it in such cases. When would the count_row be missing?

bartvanb replied:
important

JHogenboom replied:
Resolved in 3c7e72fb9d8301575a3db74dbd5bf187becc8f61


`v6-descriptive-statistics/miscellaneous.py`
```python
                if count_row.empty:
                    # No count row found - treat as privacy violation
                    privacy_violations.append(f"numerical_violation_{variable_name}")
                    continue  # Skip this variable entirely

                count_value = count_row.iloc[0]["value"]

                if count_value < sample_size_threshold:
                    # Count doesn't meet threshold - remove all statistics for this variable
                    privacy_violations.append(f"numerical_violation_{variable_name}")
                    continue  # Skip this variable entirely
```

bartvanb commented on v6-descriptive-statistics/miscellaneous.py:
```diff
@@ -0,0 +1,243 @@
+                    combined_numerical_df.to_json()
+                )
+
+    # Copy any other results that don't need threshold checking
```
This looks very risky to me. What results would this be? Do you have the full set? It would be better to have an include list instead of allowing to copy everything except the data handled explicitly above.

bartvanb replied:
important

JHogenboom replied:
Resolved in 6f81a63053d93255e8dd5f931dfd22375445c9a8 by removal of entire `for`-loop

bartvanb commented on v6-descriptive-statistics/miscellaneous.py:
```diff
@@ -0,0 +1,243 @@
+    # Only pass if all originally present data types still have valid results
+    has_statistical_results = (
+        not original_had_categorical or has_categorical_data
+    ) and (not original_had_numerical or has_numerical_data)
```
So let's say I submit 2 categorical and 1 numerical variable. If 1 categorical variable has no results, this check would succeed because I still have 1 left, right? And then, if the numerical variable failed this check would fail because none are left.
If that's correct, I think it would be better if it failed either if any variables fail, or if all fail.

bartvanb replied:
important

JHogenboom replied:
Resolved in 9f614c02b44b55bad29042f14bfb793871cb2789

`v6-descriptive-statistics/central.py`
```python
            "numerical_aggregated_results": results_general_statistics.get(
                "numerical_general_statistics", {}
            ),
```

`v6-descriptive-statistics/miscellaneous.py`
```python
    # Pass if any statistical results remain
    has_statistical_results = has_categorical_data or has_numerical_data
```

JHogenboom replied:
Added to this in https://github.com/STRONGAYA/v6-descriptive-statistics/pull/11
Reason being that a bad actor test was permitted to pass through under previous logic
```log
FAILED tests/integration/test_algorithm_integration.py::TestAlgorithmComponent::test_algorithm_inlier_specific[standard_dataset_bad_actor-central] - Failed: DID NOT RAISE Exception
FAILED tests/integration/test_algorithm_integration.py::TestAlgorithmComponent::test_algorithm_inlier_specific[standard_dataset_bad_actor-partial_general_statistics] - AssertionError: Expected one of ['CollectResultsError', 'PrivacyThresholdViolation'] but got AssertionError
```

bartvanb reviewed (commented):
First comments here as well. Let's proceed with the review after addressing this first round, mainly because I also put some questions. When I understand better, I can also do a better review.

JHogenboom committed:
Fix version dependency for vantage6-algorithm-tools

- Handling comment https://github.com/STRONGAYA/v6-descriptive-statistics/pull/9#discussion_r2571977617 in PR #9

JHogenboom committed:
use `get_env_var` default parameter instead of unnecessary `try`-block

- handling comment https://github.com/STRONGAYA/v6-descriptive-statistics/pull/9#discussion_r2584552465 in PR #9
- handling comment https://github.com/STRONGAYA/v6-descriptive-statistics/pull/9#discussion_r2584542366 in PR #9

JHogenboom committed:
update privacy threshold checks to handle missing count rows

- handling comment https://github.com/STRONGAYA/v6-descriptive-statistics/pull/9#discussion_r2609633180 in PR #9

JHogenboom committed:
remove redundant copying of results without threshold checks

- handling comment https://github.com/STRONGAYA/v6-descriptive-statistics/pull/9#discussion_r2609641903 in PR #9

JHogenboom committed:
handle missing keys in statistical results and simplify result checks

- handling comment https://github.com/STRONGAYA/v6-descriptive-statistics/pull/9#discussion_r2609653791 in PR #9

JHogenboom committed:
align versioning for test workflow

JHogenboom committed:
apply black formatting to code

JHogenboom committed:
switch to alpine image for Docker build and include `.dockerignore` for cleanliness

- vantage6 harbour is no longer in use

JHogenboom committed:
switch to alpine image for Docker build and include `.dockerignore` for cleanliness

- vantage6 harbour is no longer in use

JHogenboom committed:
update `.dockerignore`

JHogenboom committed:
switch to `pyproject.toml`

JHogenboom committed:
update `.gitignore`

JHogenboom committed:
Merge remote-tracking branch 'origin/review/important-comments' into review/important-comments

# Conflicts:
#	.dockerignore
#	Dockerfile

JHogenboom committed:
update test workflow to use `pyproject.toml` dependencies and separate linting dependencies

JHogenboom committed:
include python 3.11 and 3.12 in test environment

JHogenboom committed:
include 4.14 in testing ci

Copilot committed:
Fix inlier-specific CI test compatibility with vantage6 4.15

Copilot committed:
Stabilize inlier-specific integration expectations across backend behaviors

JHogenboom committed:
Revert "Stabilize inlier-specific integration expectations across backend behaviors"

This reverts commit 7118503dff1ed221039dcecc1cfaac4b85fa1e68.

JHogenboom committed:
Revert "Fix inlier-specific CI test compatibility with vantage6 4.15"

This reverts commit 39bfe6c436d8cada5528ec43be11caf5e1cb010b.

JHogenboom committed:
bump vantage6-strongaya-general dependency to 1.0.4

JHogenboom committed:
prepare for release

JHogenboom committed:
remove unnecessary tests for python-version matrix

JHogenboom committed:
bump vantage6-strongaya-rdf dependency to 1.0.2

Copilot committed:
fix: strict privacy threshold - raise PrivacyThresholdViolation if any variable is stripped

Replaces the permissive OR logic that allowed partial results to be
returned when some variables failed the sample size threshold. Now, if
any requested variable is stripped due to a privacy violation, the
function raises PrivacyThresholdViolation consistently for both
categorical and numerical variables.

This addresses the reviewer concern from PR #9 discussion_r2609653791
about inconsistent handling between data types, while maintaining
security against bad actors using narrow stratification.

Copilot committed:
Apply remaining changes

JHogenboom committed:
Merge pull request #11 from STRONGAYA/copilot/reviewimportant-comments

Investigating algorithm failure due to recent PR changes

Copilot committed:
Remove minimum and maximum from descriptive statistics results and documentation

Copilot committed:
Use named column accessor instead of index for robustness

Copilot committed:
Fix TypeError in adjusted deviation and tuple handling for inliers

- Add local patched compute_local_adjusted_deviation that fixes kwarg name
  mismatch (aggregated_mean vs aggregate_mean) in upstream v6-tools-general
- Add _normalize_inliers helper to convert tuple inliers to lists, fixing
  upstream strict isinstance(inliers, list) check
- Both fixes work around bugs in vantage6-strongaya-general v1.0.4

JHogenboom committed:
Merge pull request #12 from STRONGAYA/copilot/remove-min-max-results

Remove minimum and maximum from descriptive statistics results

JHogenboom committed:
refactor: reformat function signatures for improved readability; apply black

Copilot committed:
refactor: remove temporary upstream bug workarounds from partial.py

The _normalize_inliers helper, local compute_local_adjusted_deviation,
and _orchestrate_local_adjusted_deviation functions were workarounds for
bugs in vantage6-strongaya-general that have been fixed upstream in
STRONGAYA/v6-tools-general#11. Now uses the upstream functions directly.

JHogenboom committed:
Merge pull request #13 from STRONGAYA/copilot/fix-typeerror-adjusted-deviation

Remove temporary upstream bug workarounds from partial.py

JHogenboom committed:
Merge pull request #10 from STRONGAYA/review/important-comments

Review/important comments

JHogenboom commented:
Keeping this for reference and next major release
